File: core/src/rule/fluent/foundation/controls.rs
use super::super::super::shared::{append_inline_action_layout, append_transition};
use super::super::super::{declaration, token_declaration, Declaration};
use super::super::typography_from_tokens;
use crate::parse::ParsedClass;
use crate::token::TokenRegistry;

pub(super) fn resolve(parsed: &ParsedClass, tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    match (parsed.utility.as_str(), parsed.value.as_deref()) {
        ("field", None) => field(tokens),
        ("checkbox", None) => checkbox(tokens),
        ("radio", None) => radio(tokens),
        ("switch", None) => switch(tokens),
        ("textarea", None) => textarea(tokens),
        ("select", None) => select(tokens),
        ("tab", None) => tab(tokens),
        ("action", Some("primary")) => action_primary(tokens),
        ("action", Some("subtle")) => action_subtle(tokens),
        _ => None,
    }
}

pub(crate) fn field(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = typography_from_tokens(tokens, "body-size", "body-weight", None)?;
    declarations.extend([
        declaration("min-height", "2.75rem"),
        token_declaration("color", tokens.fluent.color.get("text")?),
        token_declaration("caret-color", tokens.fluent.color.get("primary")?),
        token_declaration("background-color", tokens.fluent.color.get("field")?),
        token_declaration("background-image", tokens.fluent.effect.get("field-tint")?),
        token_declaration("border", tokens.fluent.border.get("field")?),
        token_declaration("border-radius", tokens.fluent.radius.get("sm")?),
        token_declaration("padding", tokens.fluent.space.get("field-pad")?),
        token_declaration("box-shadow", tokens.fluent.shadow.get("field")?),
        declaration("backdrop-filter", "blur(14px) saturate(1.05)"),
    ]);
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}

fn checkbox(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = vec![
        declaration("appearance", "none"),
        declaration("display", "inline-grid"),
        declaration("place-items", "center"),
        declaration("inline-size", "1.125rem"),
        declaration("block-size", "1.125rem"),
        token_declaration("background-color", tokens.fluent.color.get("surface-alt")?),
        token_declaration("border", tokens.fluent.border.get("field")?),
        token_declaration("border-radius", tokens.fluent.radius.get("sm")?),
        token_declaration("box-shadow", tokens.fluent.shadow.get("field")?),
    ];
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}

fn radio(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = checkbox(tokens)?;
    declarations.retain(|decl| decl.property != "border-radius");
    declarations.push(declaration("border-radius", "999px"));
    Some(declarations)
}

fn switch(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = vec![
        declaration("appearance", "none"),
        declaration("display", "inline-flex"),
        declaration("align-items", "center"),
        declaration("inline-size", "2.5rem"),
        declaration("block-size", "1.5rem"),
        token_declaration("background-color", tokens.fluent.color.get("surface-alt")?),
        token_declaration("border", tokens.fluent.border.get("field")?),
        declaration("border-radius", "999px"),
        token_declaration("box-shadow", tokens.fluent.shadow.get("field")?),
    ];
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}

fn textarea(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = field(tokens)?;
    declarations.push(declaration("min-height", "6.5rem"));
    declarations.push(declaration("resize", "vertical"));
    Some(declarations)
}

fn select(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = field(tokens)?;
    declarations.push(declaration("appearance", "none"));
    declarations.push(declaration("background-repeat", "no-repeat"));
    declarations.push(declaration("background-position", "right 0.85rem center"));
    Some(declarations)
}

fn tab(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = typography_from_tokens(tokens, "label-size", "label-weight", None)?;
    declarations.extend([
        declaration("display", "inline-flex"),
        declaration("align-items", "center"),
        declaration("justify-content", "center"),
        declaration("min-height", "2.5rem"),
        token_declaration("color", tokens.fluent.color.get("muted")?),
        declaration("background-color", "transparent"),
        token_declaration("border-radius", tokens.fluent.radius.get("md")?),
        token_declaration("padding", tokens.fluent.space.get("action-pad")?),
        token_declaration("border", tokens.fluent.border.get("action-subtle")?),
    ]);
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}

fn action_primary(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = typography_from_tokens(tokens, "label-size", "label-weight", None)?;
    append_inline_action_layout(&mut declarations);
    declarations.extend([
        token_declaration("color", tokens.fluent.color.get("action-foreground")?),
        token_declaration("background-color", tokens.fluent.color.get("primary")?),
        token_declaration("border", tokens.fluent.border.get("action-primary")?),
        token_declaration("border-radius", tokens.fluent.radius.get("md")?),
        token_declaration("padding", tokens.fluent.space.get("action-pad")?),
        token_declaration("box-shadow", tokens.fluent.shadow.get("action")?),
    ]);
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}

fn action_subtle(tokens: &TokenRegistry) -> Option<Vec<Declaration>> {
    let mut declarations = typography_from_tokens(tokens, "label-size", "label-weight", None)?;
    append_inline_action_layout(&mut declarations);
    declarations.extend([
        token_declaration("color", tokens.fluent.color.get("primary")?),
        token_declaration(
            "background-color",
            tokens.fluent.color.get("action-subtle")?,
        ),
        token_declaration("border", tokens.fluent.border.get("action-subtle")?),
        token_declaration("border-radius", tokens.fluent.radius.get("md")?),
        token_declaration("padding", tokens.fluent.space.get("action-pad")?),
        token_declaration("box-shadow", tokens.fluent.shadow.get("action-subtle")?),
        declaration("backdrop-filter", "blur(18px) saturate(1.08)"),
    ]);
    append_transition(
        &mut declarations,
        tokens.fluent.effect.get("interactive-transition")?,
        tokens.fluent.motion.get("duration")?,
        tokens.fluent.motion.get("easing")?,
    );
    Some(declarations)
}

File: packages/motif-vite/src/core/fluent/foundation/controls.ts
import type { Declaration, ParsedClass } from '../../shared.js';
import { decl, fluent, fluentTypography, tokenDecl, withTransition } from '../../shared.js';

export function resolveFluentFoundationControls(parsed: ParsedClass): Declaration[] | null {
  switch (`${parsed.utility}:${parsed.value ?? ''}`) {
    case 'field:':
      return withTransition([
        ...fluentTypography('body-size', 'body-weight'),
        decl('min-height', '2.75rem'),
        tokenDecl('color', fluent('color', 'text')),
        tokenDecl('caret-color', fluent('color', 'primary')),
        tokenDecl('background-color', fluent('color', 'field')),
        tokenDecl('background-image', fluent('effect', 'field-tint')),
        tokenDecl('border', fluent('border', 'field')),
        tokenDecl('border-radius', fluent('radius', 'sm')),
        tokenDecl('padding', fluent('space', 'field-pad')),
        tokenDecl('box-shadow', fluent('shadow', 'field')),
        decl('backdrop-filter', 'blur(14px) saturate(1.05)'),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    case 'checkbox:':
      return withTransition([
        decl('appearance', 'none'), decl('display', 'inline-grid'), decl('place-items', 'center'), decl('inline-size', '1.125rem'), decl('block-size', '1.125rem'),
        tokenDecl('background-color', fluent('color', 'surface-alt')), tokenDecl('border', fluent('border', 'field')), tokenDecl('border-radius', fluent('radius', 'sm')), tokenDecl('box-shadow', fluent('shadow', 'field')),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    case 'radio:':
      return withTransition([
        decl('appearance', 'none'), decl('display', 'inline-grid'), decl('place-items', 'center'), decl('inline-size', '1.125rem'), decl('block-size', '1.125rem'),
        tokenDecl('background-color', fluent('color', 'surface-alt')), tokenDecl('border', fluent('border', 'field')), decl('border-radius', '999px'), tokenDecl('box-shadow', fluent('shadow', 'field')),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    case 'switch:':
      return withTransition([
        decl('appearance', 'none'), decl('display', 'inline-flex'), decl('align-items', 'center'), decl('inline-size', '2.5rem'), decl('block-size', '1.5rem'),
        tokenDecl('background-color', fluent('color', 'surface-alt')), tokenDecl('border', fluent('border', 'field')), decl('border-radius', '999px'), tokenDecl('box-shadow', fluent('shadow', 'field')),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    case 'textarea:':
      return withTransition([
        ...fluentTypography('body-size', 'body-weight'), decl('min-height', '6.5rem'), decl('resize', 'vertical'), tokenDecl('color', fluent('color', 'text')), tokenDecl('caret-color', fluent('color', 'primary')),
        tokenDecl('background-color', fluent('color', 'field')), tokenDecl('background-image', fluent('effect', 'field-tint')), tokenDecl('border', fluent('border', 'field')), tokenDecl('border-radius', fluent('radius', 'sm')), tokenDecl('padding', fluent('space', 'field-pad')), tokenDecl('box-shadow', fluent('shadow', 'field')), decl('backdrop-filter', 'blur(14px) saturate(1.05)'),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    case 'select:':
      return withTransition([
        ...fluentTypography('body-size', 'body-weight'), decl('min-height', '2.75rem'), decl('appearance', 'none'), decl('background-repeat', 'no-repeat'), decl('background-position', 'right 0.85rem center'), tokenDecl('color', fluent('color', 'text')),
        tokenDecl('background-color', fluent('color', 'field')), tokenDecl('background-image', fluent('effect', 'field-tint')), tokenDecl('border', fluent('border', 'field')), tokenDecl('border-radius', fluent('radius', 'sm')), tokenDecl('padding', fluent('space', 'field-pad')), tokenDecl('box-shadow', fluent('shadow', 'field')),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    case 'tab:':
      return withTransition([
        ...fluentTypography('label-size', 'label-weight'), decl('display', 'inline-flex'), decl('align-items', 'center'), decl('justify-content', 'center'), decl('min-height', '2.5rem'), tokenDecl('color', fluent('color', 'muted')),
        decl('background-color', 'transparent'), tokenDecl('border-radius', fluent('radius', 'md')), tokenDecl('padding', fluent('space', 'action-pad')), tokenDecl('border', fluent('border', 'action-subtle')),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    case 'dialog:':
      return withTransition([
        tokenDecl('background-color', fluent('color', 'panel')), tokenDecl('background-image', fluent('effect', 'panel-tint')), decl('background-blend-mode', 'screen'), tokenDecl('color', fluent('color', 'text')), tokenDecl('border', fluent('border', 'panel')),
        tokenDecl('border-radius', fluent('radius', 'lg')), tokenDecl('padding', fluent('space', 'panel-pad')), tokenDecl('box-shadow', fluent('shadow', 'panel')), decl('backdrop-filter', 'blur(24px) saturate(1.2)'), decl('max-width', '32rem'), decl('width', '100%'),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    case 'list:item':
      return withTransition([
        decl('display', 'flex'), decl('align-items', 'center'), decl('justify-content', 'space-between'), decl('gap', '0.75rem'), decl('min-height', '3rem'), tokenDecl('color', fluent('color', 'text')), tokenDecl('background-color', fluent('color', 'surface-alt')),
        tokenDecl('background-image', fluent('effect', 'surface-alt-tint')), decl('background-blend-mode', 'screen'), tokenDecl('border', fluent('border', 'action-subtle')), tokenDecl('border-radius', fluent('radius', 'md')), tokenDecl('padding', fluent('space', 'surface-pad-sm')), tokenDecl('box-shadow', fluent('shadow', 'surface-alt')), decl('backdrop-filter', 'blur(18px) saturate(1.08)'),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    case 'menu:item':
      return withTransition([
        decl('display', 'flex'), decl('align-items', 'center'), decl('justify-content', 'space-between'), decl('gap', '0.75rem'), decl('min-height', '2.5rem'), decl('cursor', 'pointer'), tokenDecl('color', fluent('color', 'text')), tokenDecl('background-color', fluent('color', 'surface-alt')),
        tokenDecl('background-image', fluent('effect', 'surface-alt-tint')), decl('background-blend-mode', 'screen'), tokenDecl('border', fluent('border', 'action-subtle')), tokenDecl('border-radius', fluent('radius', 'md')), tokenDecl('padding', fluent('space', 'surface-pad-sm')), tokenDecl('box-shadow', fluent('shadow', 'surface-alt')), decl('backdrop-filter', 'blur(18px) saturate(1.08)'),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    case 'icon:button':
      return withTransition([
        decl('display', 'inline-grid'), decl('place-items', 'center'), decl('inline-size', '2.5rem'), decl('block-size', '2.5rem'), tokenDecl('color', fluent('color', 'primary')), tokenDecl('background-color', fluent('color', 'action-subtle')),
        tokenDecl('border', fluent('border', 'action-subtle')), tokenDecl('border-radius', fluent('radius', 'md')), tokenDecl('box-shadow', fluent('shadow', 'action-subtle')),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    case 'nav:item':
      return withTransition([
        ...fluentTypography('label-size', 'label-weight'), decl('display', 'inline-flex'), decl('align-items', 'center'), decl('gap', '0.5rem'), decl('min-height', '2.5rem'), tokenDecl('color', fluent('color', 'text')),
        tokenDecl('background-color', fluent('color', 'action-subtle')), tokenDecl('border', fluent('border', 'action-subtle')), tokenDecl('border-radius', fluent('radius', 'md')), tokenDecl('padding', fluent('space', 'action-pad')), tokenDecl('box-shadow', fluent('shadow', 'action-subtle')),
      ], fluent('effect', 'interactive-transition'), fluent('motion', 'duration'), fluent('motion', 'easing'));
    default:
      return null;
  }
}
